### PR TITLE
kubelet: fix potential resource leak

### PR DESCRIPTION
**Pull request by xyz-li**
#### What type of PR is this?

/kind flake

#### What this PR does / why we need it:
A sample code mimic the function `ReadLogs`
It may cause resource leaks and close the file twice.
```
func closeFileTwice() {
	fn := "/tmp/abc.txt"
	f, _ := os.Open(fn)
	defer f.Close()

	for i := 0; i < 10; i++ {
		newF, _ := os.Open(fn)
		defer func(f *os.File) {
			if err := f.Close(); err != nil {
				panic(err)
			}
		}(newF)

		f.Close()
		f = newF
	}
}

```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 1]
```diff
@@ -295,7 +295,8 @@ func ReadLogs(ctx context.Context, path, containerID string, opts *LogOptions, r
 	if err != nil {
 		return fmt.Errorf("failed to open log file %q: %v", path, err)
 	}
-	defer f.Close()
+	// Let the closure capture the variable f, so it can close the last opened file.
+	defer func() { f.Close() }()
```
Comment (aojea) [inline pkg/kubelet/kuberuntime/logs/logs.go:299]: f is not the variable in the loop , you don't have to capture it

This behavior will also be fixed in golang https://github.com/golang/go/discussions/56010

Comment (bart0sh) [reply pkg/kubelet/kuberuntime/logs/logs.go]: I agree with @xyz-li here. We're at the risk of double close and this PR fixes the issue.

However, I'm a bit concerned about updating `f` variable in the `ReadLogs` function.  It doesn't look clear and readable to me.
Would it make sense to wrap all that logic into a function and pass `f` as a parameter there?

/reopen

Comment (bart0sh) [reply pkg/kubelet/kuberuntime/logs/logs.go]: @giuseppe @aojea WDYT?

Comment (aojea) [reply pkg/kubelet/kuberuntime/logs/logs.go]: well, my understanding of defer was different https://go.dev/ref/spec#Defer_statements as f.Close() executed will be the actual f evaluated at the beginning of the function ... but I may be wrong 

independently of that, +1 for what Ed says, let's make this easier to read and avoid to need this deep knowledge of the language

Comment (bart0sh) [reply pkg/kubelet/kuberuntime/logs/logs.go]: @aojea to my understanding (docs you've referred suggest the same) deferred function value is avaluated at the time of deferring (not executing).

Here is a simple code demonstrating this behaviour: https://go.dev/play/p/H12eFAXbi89

**Comment by bart0sh** [comment]
/priority important-longterm
/triage accepted

@xyz-li please address review comments or close the PR if you're not willing to continue.

**Comment by xyz-li** [comment]
> /priority important-longterm /triage accepted
> 
> @xyz-li please address review comments or close the PR if you're not willing to continue.

I will address it in 2-3 days.

**Comment by xyz-li** [comment]
@bart0sh 
How about this commit h #124212?

**Comment by bart0sh** [comment]
@xyz-li 
> How about this commit h #124212?

This is what I think about it: https://github.com/kubernetes/kubernetes/pull/121377#discussion_r1549556378
And this is what @aojea thinks: https://github.com/kubernetes/kubernetes/pull/121377#discussion_r1549627929

Both comments have to be addressed before we can move forward with this PR.

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 2]
```diff
@@ -367,9 +391,11 @@ func ReadLogs(ctx context.Context, path, containerID string, opts *LogOptions, r
 						}
 						return fmt.Errorf("failed to open log file %q: %v", path, err)
 					}
-					defer newF.Close()
-					f.Close()
+
 					f = newF
```
Comment (bart0sh) [inline pkg/kubelet/kuberuntime/logs/logs.go:395]: We're still changing `f` in the middle of processing, so I don't think last change simplified the code. Previous variant was at least shorter. Did you try [this suggestion](https://github.com/kubernetes/kubernetes/pull/121377#discussion_r1549556378)?

File: pkg/kubelet/kuberuntime/logs/logs.go [hunk 3]
```diff
@@ -277,6 +277,27 @@ func (w *logWriter) write(msg *logMessage, addPrefix bool) error {
...
+	if fc.oldFile != nil {
+		klog.V(4).InfoS("Closing log file", "name", fc.oldFile.Name())
+		if err := fc.oldFile.Close(); err != nil {
+			return err
+		}
+	}
+	fc.oldFile = newF
+	return nil
+}
+
```
Comment (bart0sh) [inline pkg/kubelet/kuberuntime/logs/logs.go:300]: This is better, but still I think the code is hard to read.
I'd propose to return back to the original short bugfix PR and do a refactoring work in a separate PR. 
Does this make sense?

Comment (xyz-li) [reply pkg/kubelet/kuberuntime/logs/logs.go:300]: Do you mean this commit? https://github.com/xyz-li/kubernetes/commit/2ff8b7b52ce3af7312c8a97707a9614e7847495c

Then we must open the new log file, and close the previous opened file, assign newF to the variable `f`.
https://github.com/xyz-li/kubernetes/commit/2ff8b7b52ce3af7312c8a97707a9614e7847495c?diff=split&w=0#diff-5c455ea60fb38862bb012a440688cb9f8319e5e5895ce5599038037689120f05R375

I haven't found a method to eliminate the assignment `f = newF`.

Comment (bart0sh) [reply pkg/kubelet/kuberuntime/logs/logs.go:300]: > Then we must open the new log file, and close the previous opened file, assign newF to the variable f.
https://github.com/xyz-li/kubernetes/commit/059d2c5abedb60ba0ae2d5c6f23683f2c1d4197d#diff-5c455ea60fb38862bb012a440688cb9f8319e5e5895ce5599038037689120f05R376-R377

yes, please do. That should fix the issue.  
And then, if you want to continue, you can refactor the code in a separate PR.

Comment (xyz-li) [reply pkg/kubelet/kuberuntime/logs/logs.go:300]: #124298. This is actually the initial commit.

Comment (bart0sh) [reply pkg/kubelet/kuberuntime/logs/logs.go:300]: Thanks. I'll lgtm it if/when CI is green.
